subsys/fs/littlefs_fs.c

```diff
@@ -541,154 +541,236 @@ static bool get_page_cb(const struct flash_pages_info *info, void *ctxp)
+	BUILD_ASSERT(CONFIG_FS_LITTLEFS_LOOKAHEAD_SIZE > 0);
+	BUILD_ASSERT((CONFIG_FS_LITTLEFS_LOOKAHEAD_SIZE % 8) == 0);
+	BUILD_ASSERT((CONFIG_FS_LITTLEFS_CACHE_SIZE % CONFIG_FS_LITTLEFS_READ_SIZE) == 0);
+	BUILD_ASSERT((CONFIG_FS_LITTLEFS_CACHE_SIZE % CONFIG_FS_LITTLEFS_PROG_SIZE) == 0);
+
+	size_t device_size = device_get_size(lcp->context);
+
+	if (lcp->read_size == 0) {
+		lcp->read_size = CONFIG_FS_LITTLEFS_READ_SIZE;
 	}
 
-	/* Create and take mutex. */
-	k_mutex_init(&fs->mutex);
-	fs_lock(fs);
+	if (lcp->prog_size == 0) {
+		lcp->prog_size = CONFIG_FS_LITTLEFS_PROG_SIZE;
+	}
 
-	/* Open flash area */
-	ret = flash_area_open(area_id, &fs->area);
-	if ((ret < 0) || (fs->area == NULL)) {
-		LOG_ERR("can't open flash area %d", area_id);
-		ret = -ENODEV;
-		goto out;
+	if (lcp->block_size == 0) {
+		lcp->block_size = device_get_block_size(lcp->context);
 	}
-	LOG_DBG("FS area %u at 0x%x for %u bytes",
-		area_id, (uint32_t)fs->area->fa_off,
-		(uint32_t)fs->area->fa_size);
 
-	dev = flash_area_get_device(fs->area);
-	if (dev == NULL) {
-		LOG_ERR("can't get flash device: %s", log_strdup(fs->area->fa_dev_name));
-		ret = -ENODEV;
-		goto out;
+	if (lcp->block_size == 0) {
+		__ASSERT_NO_MSG(lcp->block_size != 0);
+		return -EINVAL;
 	}
 
-	BUILD_ASSERT(CONFIG_FS_LITTLEFS_READ_SIZE > 0);
-	BUILD_ASSERT(CONFIG_FS_LITTLEFS_PROG_SIZE > 0);
-	BUILD_ASSERT(CONFIG_FS_LITTLEFS_CACHE_SIZE > 0);
-	BUILD_ASSERT(CONFIG_FS_LITTLEFS_LOOKAHEAD_SIZE > 0);
-	BUILD_ASSERT((CONFIG_FS_LITTLEFS_LOOKAHEAD_SIZE % 8) == 0);
-	BUILD_ASSERT((CONFIG_FS_LITTLEFS_CACHE_SIZE
-		      % CONFIG_FS_LITTLEFS_READ_SIZE) == 0);
-	BUILD_ASSERT((CONFIG_FS_LITTLEFS_CACHE_SIZE
-		      % CONFIG_FS_LITTLEFS_PROG_SIZE) == 0);
+	if (lcp->block_cycles == 0) {
+		lcp->block_cycles = CONFIG_FS_LITTLEFS_BLOCK_CYCLES;
+	}
 
-	struct lfs_config *lcp = &fs->cfg;
+	if (lcp->block_cycles <= 0) {
+		/* Disable leveling (littlefs v2.1+ semantics) */
+		lcp->block_cycles = -1;
+	}
 
-	lfs_size_t read_size = lcp->read_size;
+	if (lcp->cache_size == 0) {
+		lcp->cache_size = CONFIG_FS_LITTLEFS_CACHE_SIZE;
+	}
 
-	if (read_size == 0) {
-		read_size = CONFIG_FS_LITTLEFS_READ_SIZE;
+	if (lcp->lookahead_size == 0) {
+		lcp->lookahead_size = CONFIG_FS_LITTLEFS_LOOKAHEAD_SIZE;
 	}
 
-	lfs_size_t prog_size = lcp->prog_size;
+	/* No, you don't get to override this. */
+	lcp->block_count = device_size / lcp->block_size;
+
+	__ASSERT_NO_MSG(lcp->prog_size != 0);
+	__ASSERT_NO_MSG(lcp->read_size != 0);
+	__ASSERT_NO_MSG(lcp->cache_size != 0);
+	__ASSERT_NO_MSG(lcp->block_size != 0);
+
+	__ASSERT((device_size % lcp->block_size) == 0,
+		 "partition size must be multiple of block size");
+	__ASSERT((lcp->block_size % lcp->prog_size) == 0,
+		 "program size must be multiple of write size");
+	__ASSERT((lcp->block_size % lcp->cache_size) == 0,
+		 "cache size incompatible with block size");
+
+	return 0;
+}
+
+static int lfs_config_initialize_operations(struct fs_littlefs *fs)
```

Comment:
Is this function still required?

Reply:
Yes, because previously user might have configured the fs_littlefs manually. In such a case they propably did not set the read, prog, erase and sync functions. Hence, we still have to set the defaults in case no operation at all has been set.

Reply:
Assigning a default based upon NULL makes understanding the code difficult and unpredictable. To avoid any problems and catching all errors (even when the function pointers are changed during execution) a method could be to add the read/write/... functions to your device_services (these could be const) and to define fs->cfg->read/... functions inside littlefs_fs.c:
```
.cfg.read = little_fs_read;
...
int little_fs_read(...)
{
  if (!fs.device_services->read) {
    return -ENOTSUP;
  }
  return fs.devices_services->read(...);
}
```

Reply:
I don't think it is possible to make devices_services const, as its members might already end up not being set. In case somebody created the instance of fs_littlefs manually. But I agree with the rest, it would be more clear to bundle these functions together in one spot.